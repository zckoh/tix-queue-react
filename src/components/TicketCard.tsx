import moment from "moment";
import { Row, Col, Card, Button } from "react-bootstrap";

import { Ticket } from "../types";

const prepareCardDescription = (description: String) => {
  if (description.length >= 150) return description.substring(1, 150) + `...`;
  return description;
};

interface TicketCardPropTypes {
  onClickViewTicket: (ticket: Ticket) => void;
  ticket: Ticket;
}

const TicketCard = (props: TicketCardPropTypes) => {
  const { onClickViewTicket, ticket } = props;

  return (
    <Card bg="dark" className="my-3">
      <Card.Body>
        <Card.Title className="text-truncate">{ticket.title}</Card.Title>
        <Card.Text>{prepareCardDescription(ticket.description)}</Card.Text>
      </Card.Body>
      <Card.Footer>
        <Row>
          <Col>
            <small className="text-muted">
              Created <br /> {moment(ticket.createdAt).fromNow()}
            </small>
          </Col>
          <Col>
            <Button variant="primary" className="float-right" onClick={() => onClickViewTicket(ticket)}>
              View Ticket
            </Button>
          </Col>
        </Row>
      </Card.Footer>
    </Card>
  );
};

export default TicketCard;
