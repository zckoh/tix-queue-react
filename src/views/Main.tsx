import { useCallback, useEffect, useState } from "react";
import { Container, Row, Col, Button, Toast, Dropdown, DropdownButton } from "react-bootstrap";
import { v4 as uuidv4 } from "uuid";
import moment from "moment";

import TicketCard from "../components/TicketCard";
import TicketDetailsModal from "../components/TicketDetailsModal";
import TicketCreateModal from "../components/TicketCreateModal";
import { getAllDepartments, getAllStatus, getAllTechnicians, getAllTickets } from "../services";
import { PropTypes, Status, Ticket, Notification, Department, Technician } from "../types";

import "../styles/Main.css";
import "../styles/Toast.css";

const emptyTicket: Ticket = {
  id: undefined,
  title: "",
  author: "",
  description: "",
  status: undefined,
  createdAt: "",
  solution: null,
  department: undefined,
  technician: undefined,
};

const Main = (props: PropTypes) => {
  const [showTicketModal, setShowTicketModal] = useState(false);
  const [showCreateModal, setShowCreateModal] = useState(false);

  // States populated from backend
  const [allTickets, setAllTickets] = useState<Ticket[]>([]);
  const [allStatus, setAllStatus] = useState<Status[]>([]);
  const [allDepartments, setAllDepartments] = useState<Department[]>([]);
  const [allTechnicians, setAllTechnicians] = useState<Technician[]>([]);

  // States populated from frontend
  const [currentTicket, setCurrentTicket] = useState<Ticket>(emptyTicket);
  const [notifications, setNotifications] = useState<Notification[]>([]);

  const hideNotification = (uuid: String) => {
    setNotifications((currentNotifications) => {
      const notificationIndex = currentNotifications.findIndex((n) => n.notificationId === uuid);
      let updatedNotifications = [...currentNotifications];

      updatedNotifications[notificationIndex].show = false;

      // clear notifications array if it gets too large
      if (updatedNotifications.length > 10) {
        updatedNotifications = [];
      }
      return updatedNotifications;
    });
  };

  const closeToast = (uuid: String) => () => hideNotification(uuid);

  const fetchAllStatus = useCallback(async () => {
    getAllStatus()
      .catch((error) => {
        updateNotificationList({
          notificationId: uuidv4(),
          message: "❌ - Failed to load all status.",
          status: "failed",
          show: true,
          createdAt: moment(),
        });
        console.error(error);
        throw error;
      })
      .then((statuses) => {
        setAllStatus(statuses);
      });
  }, []);

  const fetchAllDepartments = useCallback(async () => {
    getAllDepartments()
      .catch((error) => {
        updateNotificationList({
          notificationId: uuidv4(),
          message: "❌ - Failed to load all departments.",
          status: "failed",
          show: true,
          createdAt: moment(),
        });
        console.error(error);
        throw error;
      })
      .then((departments) => {
        setAllDepartments(departments);
      });
  }, []);

  const fetchAllTechnicians = useCallback(async () => {
    await getAllTechnicians()
      .catch((error) => {
        updateNotificationList({
          notificationId: uuidv4(),
          message: "❌ - Failed to load all technicians.",
          status: "failed",
          show: true,
          createdAt: moment(),
        });
        console.error(error);
        throw error;
      })
      .then((technician) => {
        setAllTechnicians(technician);
      });
  }, []);

  const fetchAllTickets = useCallback(async () => {
    await getAllTickets()
      .catch((error) => {
        updateNotificationList({
          notificationId: uuidv4(),
          message: "❌ - Failed to load all tickets.",
          status: "failed",
          show: true,
          createdAt: moment(),
        });
        console.error(error);
        throw error;
      })
      .then((tickets) => {
        setAllTickets(tickets);
      });
  }, []);

  useEffect(() => {
    fetchAllStatus();
    fetchAllDepartments();
    fetchAllTickets();
    fetchAllTechnicians();
  }, [fetchAllTickets, fetchAllStatus, fetchAllDepartments, fetchAllTechnicians]);

  const updateNotificationList = (newNotification: Notification): void => {
    setNotifications((currentNotifications) => [...currentNotifications, newNotification]);
  };

  const handleCloseTicketModal = () => {
    setShowTicketModal(false);
    setCurrentTicket(emptyTicket);
    fetchAllTickets();
  };
  const handleShowTicketModal = (ticket: Ticket) => {
    setCurrentTicket(ticket);
    setShowTicketModal(true);
  };

  const handleCloseCreateModal = () => {
    setShowCreateModal(false);
    fetchAllTickets();
  };
  const handleShowCreateModal = () => setShowCreateModal(true);

  const refreshAll = async () => {
    let success = true;
    try {
      await fetchAllTickets();
      await fetchAllStatus();
      await fetchAllDepartments();
      await fetchAllTechnicians();
    } catch (error) {
      success = false;
    }

    if (success) {
      updateNotificationList({
        notificationId: uuidv4(),
        message: "✔️ - Successfully refreshed all tickets.",
        status: "success",
        show: true,
        createdAt: moment(),
      });
    }
  };

  const [sortOption, setSortOption] = useState("0");

  const displaySortByTitle = () => {
    switch (sortOption) {
      case "1":
        return "Oldest to Newest";
      case "0":
        return "Newest to Oldest";
      default:
        return "Oldest to Newest";
    }
  };

  const selectSortBy = (eventKey: string | null) => {
    eventKey && setSortOption(eventKey);
  };

  const [filterOption, setFilterOption] = useState<String>("0");

  const displayFilterByTitle = () => {
    if (filterOption === "0") return `All Departments`;
    if (filterOption === "N/A") return filterOption;
    return allDepartments?.find((dep) => dep.id.toString() === filterOption)?.name;
  };
  const selectFilterBy = (eventKey: string | null) => {
    eventKey && setFilterOption(eventKey);
  };

  const allTicketsSorter = (a: any, b: any) => {
    if (sortOption === "0") return -moment(a.createdAt).diff(b.createdAt);
    return moment(a.createdAt).diff(b.createdAt);
  };

  return (
    <main className={`Main ${props.className}`}>
      <Container>
        <Row className="py-5">
          <Col className="d-flex align-items-end">
            <Button variant="success" onClick={handleShowCreateModal}>
              Create new ticket
            </Button>
            <Button variant="secondary" className="ml-3" onClick={refreshAll}>
              Refresh
            </Button>
          </Col>
          <Col>
            <Row>
              <Col lg={8}>
                <p className="text-right">Sort By:</p>
                <DropdownButton
                  variant="secondary"
                  className="float-right"
                  menuAlign="right"
                  title={displaySortByTitle()}
                  onSelect={selectSortBy}
                >
                  <Dropdown.Item eventKey="0">Newest to Oldest</Dropdown.Item>
                  <Dropdown.Item eventKey="1">Oldest to Newest</Dropdown.Item>
                </DropdownButton>
              </Col>
              {/* <Col lg={4}>
                <p className="text-right">Filter By Author:</p>
                <DropdownButton
                  className="float-right"
                  variant="info"
                  menuAlign="right"
                  title={displayFilterByTitle()}
                  onSelect={selectFilterBy}
                >
                  <Dropdown.Item eventKey="0">All</Dropdown.Item>
                  {allTechnicians?.map((department, i) => (
                    <Dropdown.Item eventKey={department.id.toString()} key={i}>
                      {department.name}
                    </Dropdown.Item>
                  ))}
                </DropdownButton>
              </Col> */}

              <Col lg={4}>
                <p className="text-right">Filter By Department:</p>
                <DropdownButton
                  className="float-right"
                  menuAlign="right"
                  title={displayFilterByTitle()}
                  onSelect={selectFilterBy}
                >
                  <Dropdown.Item eventKey="0">All Departments</Dropdown.Item>
                  {allDepartments?.map((department, i) => (
                    <Dropdown.Item eventKey={department.id.toString()} key={i}>
                      {department.name}
                    </Dropdown.Item>
                  ))}
                  <Dropdown.Item eventKey="N/A">N/A</Dropdown.Item>
                </DropdownButton>
              </Col>
            </Row>
          </Col>
        </Row>
        <Row>
          {allStatus?.map((status, i) => (
            <Col lg={Math.floor(12 / allStatus.length)} key={i}>
              <h2>{status.name}</h2>
              {allTickets
                ?.filter((ticket) => {
                  if (filterOption === "0") return true;
                  if (filterOption === "N/A") return ticket.department === null;
                  return ticket.department?.id.toString() === filterOption;
                })
                ?.filter((ticket) => ticket.status?.name === status.name)
                ?.sort(allTicketsSorter)
                .map((ticket, i) => (
                  <TicketCard ticket={ticket} onClickViewTicket={handleShowTicketModal} key={i} />
                ))}
            </Col>
          ))}
        </Row>
      </Container>
      <TicketDetailsModal
        show={showTicketModal}
        ticket={currentTicket}
        allStatus={allStatus}
        allDepartments={allDepartments}
        allTechnicians={allTechnicians}
        handleClose={handleCloseTicketModal}
        handleNotification={updateNotificationList}
      />
      <TicketCreateModal
        show={showCreateModal}
        allStatus={allStatus}
        allDepartments={allDepartments}
        allTechnicians={allTechnicians}
        handleClose={handleCloseCreateModal}
        handleNotification={updateNotificationList}
      />
      <div
        style={{
          position: "fixed",
          top: 10,
          right: 10,
          zIndex: 1051,
          width: `250px`,
        }}
      >
        {notifications?.map((notification, i) => (
          <Toast
            key={i}
            onClose={closeToast(notification.notificationId)}
            show={notification.show}
            delay={3000}
            autohide
            className={`toast-${notification.status}`}
          >
            <Toast.Header>
              <strong className="mr-auto">Notification</strong>
              <small>{moment(notification.createdAt).fromNow()}</small>
            </Toast.Header>
            <Toast.Body>{notification.message}</Toast.Body>
          </Toast>
        ))}
      </div>
    </main>
  );
};

export default Main;
