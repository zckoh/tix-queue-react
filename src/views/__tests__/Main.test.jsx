import { mount } from "enzyme";
import { Button, Modal, Toast } from "react-bootstrap";
import { waitFor } from "@testing-library/react";
import Main from "../../views/Main";
import { act, render, cleanup } from "@testing-library/react";
import { enableFetchMocks } from "jest-fetch-mock";
import TicketCard from "../../components/TicketCard";
enableFetchMocks();

beforeEach(() => {
  fetch.resetMocks();
  // hide console.error when testing
  jest.spyOn(console, "error").mockImplementation(() => {});
});

const allStatus = [
  {
    id: 1,
    name: "New",
  },
  {
    id: 2,
    name: "In Progress",
  },
  {
    id: 3,
    name: "Completed",
  },
];

const allTechnicians = [
  {
    id: 4,
    name: "Zac Koh",
    department: {
      id: 2,
      name: "IT",
    },
  },
  {
    id: 6,
    name: "Sam smith",
    department: {
      id: 3,
      name: "Cloud",
    },
  },
  {
    id: 7,
    name: "post.malone@helpqueue.com",
    department: {
      id: 3,
      name: "Cloud",
    },
  },
];

const allDepartments = [
  {
    id: 1,
    name: "Hardware",
  },
  {
    id: 2,
    name: "IT",
  },
  {
    id: 3,
    name: "Cloud",
  },
  {
    id: 9,
    name: "AWS",
  },
];

const mockAllTicketsData = [
  {
    id: 1,
    title: "whatt",
    author: "test@test.com",
    description: "updated description",
    solution: "aweosme solution",
    createdAt: "2021-04-05T12:20:01.939398",
    status: {
      id: 3,
      name: "Completed",
    },
    department: null,
    technician: null,
  },
  {
    id: 3,
    title: "New ticket",
    author: "new@we.com",
    description: "no solution yet",
    solution: null,
    createdAt: "2021-04-07T21:53:59.768904",
    status: {
      id: 2,
      name: "In Progress",
    },
    department: {
      id: 1,
      name: "Hardware",
    },
    technician: null,
  },
  {
    id: 5,
    title: "Tick without department",
    author: "zhengcong.koh@barclays.com",
    description: "THis is the second 22222 asdsad the ticket",
    solution: null,
    createdAt: "2021-04-09T14:52:33.554073",
    status: {
      id: 2,
      name: "In Progress",
    },
    department: null,
    technician: null,
  },
  {
    id: 6,
    title: "Tick with department",
    author: "zhengcong.koh@barclays.com",
    description: "THis is the second 22222 asdsad the ticket",
    solution: null,
    createdAt: "2021-04-09T14:53:09.408862",
    status: {
      id: 2,
      name: "In Progress",
    },
    department: null,
    technician: null,
  },
  {
    id: 7,
    title: "Tick with department",
    author: "zhengcong.koh@barclays.com",
    description: "THis is the second 22222 asdsad the ticket",
    solution: null,
    createdAt: "2021-04-09T14:53:25.262207",
    status: {
      id: 3,
      name: "Completed",
    },
    department: {
      id: 2,
      name: "IT",
    },
    technician: null,
  },
  {
    id: 8,
    title: "Hardware ticket",
    author: "ticket@hardware.com",
    description: "laksjd",
    solution: "done solution",
    createdAt: "2021-04-12T21:37:58.662237",
    status: {
      id: 3,
      name: "Completed",
    },
    department: {
      id: 3,
      name: "Cloud",
    },
    technician: null,
  },
  {
    id: 9,
    title: "Add more test cases to our application",
    author: "nothgin@not.com",
    description: "Adding more test cases will help improve our code quality",
    solution: "",
    createdAt: "2021-04-12T21:43:31.636341",
    status: {
      id: 3,
      name: "Completed",
    },
    department: null,
    technician: {
      id: 4,
      name: "Zac Koh",
      department: {
        id: 2,
        name: "IT",
      },
    },
  },
  {
    id: 10,
    title: "ticket without asignee & department",
    author: "test@test.com",
    description: "lkwq",
    solution: null,
    createdAt: "2021-04-14T15:12:41.983046",
    status: {
      id: 1,
      name: "New",
    },
    department: {
      id: 1,
      name: "Hardware",
    },
    technician: {
      id: 5,
      name: "Tif Koh",
      department: {
        id: 2,
        name: "IT",
      },
    },
  },
  {
    id: 11,
    title: "ticket with department & technician",
    author: "qlkwe@lqkwe.com",
    description: "lkasj",
    solution: null,
    createdAt: "2021-04-14T15:13:10.649452",
    status: {
      id: 1,
      name: "New",
    },
    department: {
      id: 2,
      name: "IT",
    },
    technician: {
      id: 4,
      name: "Zac Koh",
      department: {
        id: 2,
        name: "IT",
      },
    },
  },
  {
    id: 14,
    title: "new fixed ticket",
    author: "ticket@fixe.com",
    description: "lkasdj",
    solution: null,
    createdAt: "2021-04-14T15:20:43.993831",
    status: {
      id: 1,
      name: "New",
    },
    department: {
      id: 1,
      name: "Hardware",
    },
    technician: {
      id: 7,
      name: "post.malone@helpqueue.com",
      department: {
        id: 3,
        name: "Cloud",
      },
    },
  },
];

describe("<Main />", () => {
  const mockHandleClose = jest.fn();
  const mockHandleNotification = jest.fn();
  const mockHandleSuccessDelete = jest.fn();

  const inputTicketTitle = "This is a ticket title";
  const inputTicketId = 10;

  it(`should correctly render component`, async () => {
    let wrapper;
    fetch.mockResponses(
      [JSON.stringify(allStatus)],
      [JSON.stringify(allTechnicians)],
      [JSON.stringify(mockAllTicketsData)],
      [JSON.stringify(allDepartments)],
    );
    await waitFor(() => {
      wrapper = mount(<Main className="py-5" />);
    });
    expect(wrapper).toMatchSnapshot();
    expect(fetch).toHaveBeenCalledTimes(4);
  });

  it(`when clicked refresh button, should refetch data again`, async () => {
    let wrapper;
    fetch.mockResponses(
      [JSON.stringify(allStatus)],
      [JSON.stringify(allTechnicians)],
      [JSON.stringify(mockAllTicketsData)],
      [JSON.stringify(allDepartments)],
      [JSON.stringify(allStatus)],
      [JSON.stringify(allTechnicians)],
      [JSON.stringify(mockAllTicketsData)],
      [JSON.stringify(allDepartments)],
    );
    // Wait for Main component to be mounted
    await waitFor(() => {
      wrapper = mount(<Main className="py-5" />);
    });
    wrapper.update();

    const refreshButton = wrapper
      .find(Button)
      .findWhere((n) => n.prop("children") === "Refresh")
      .first();

    refreshButton.simulate("click");
    wrapper.update();

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledTimes(8);
    });
  });
});
