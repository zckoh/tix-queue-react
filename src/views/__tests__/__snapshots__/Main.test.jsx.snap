// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> should correctly render component 1`] = `ReactWrapper {}`;
